### PR TITLE
Async internals

### PR DESCRIPTION
**Pull request by wookiehangover**
#### What does it do?

This refactors to add test coverage, fix a false-positive passing test, and moves away from internal promise usage to facilitate testing.
#### Where should the reviewer start?

Sorry for the gross diff, but the bulk of the work is moving the internals of `lib/index.js` to use async instead of promises for handling paginated requests.
#### Any background context?

It's tough to see why promises can be a pain when mixing them with error-first style callbacks. This is a good explanation behind why it's bad: http://blog.izs.me/post/59142742143/designing-apis-for-asynchrony

In our case, it was making it difficult to test and mock the internals of search.

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -20,66 +27,70 @@ var getOpts = function (searchQuery, options) {
...
         }
-        d.resolve(body2);
-      });
-    });
 
-    Q.all(strs)
-      .then(function (jsonStrings) {
-        // jsonStrings is an array in the form of [XHR, jsonString]
+        if (pageResp.statusCode !== 200) {
+          return next(new Error('Unexpected status code'));
```
Comment (justinabrahms) [inline lib/index.js]: Can you include the status code in the error?

File: lib/index.js [hunk 2]
```diff
@@ -1,10 +1,17 @@
-var nodeRequest = require('request');
+var request = require('request');
 var _ = require('lodash');
 var Q = require('q');
```
Comment (justinabrahms) [inline lib/index.js:3]: Can you clean up the unused imports? (here and in tests)

File: lib/index.js (hunk 1 above)
Comment (wookiehangover) [inline lib/index.js]: :+1:

File: lib/index.js (hunk 2 above)
Comment (wookiehangover) [inline lib/index.js:3]: Q is still used to provide a promise interface.

**Comment by justinabrahms** [comment]
:shipit:

**Pull request merged**